src/Nixfmt/Predoc.hs

```diff
@@ -0,0 +1,195 @@
+    , hardline
+    , emptyline
+    , Doc
+    , Pretty
+    , pretty
+    , putDocW
+    , fixup
+    , flatten
+    ) where
+
+import           Data.Text                      (Text, pack)
+import qualified Data.Text.Prettyprint.Doc      as PP
+import qualified Data.Text.Prettyprint.Doc.Util as PPU
+import           Prelude                        hiding (break)
+
+data Tree a = EmptyTree
+            | Leaf a
+            | Node (Tree a) (Tree a)
+            deriving (Show, Functor)
+
+-- | Sequential Lines are reduced to a single Line by taking the maximum. This
+-- means that e.g. a Space followed by an Emptyline results in just an
+-- Emptyline.
+data Line = Softbreak
+          | Break
+          | Softline
+          | Space
+          | Hardline
+          | Emptyline
+          deriving (Show, Eq, Ord)
+
+data Predoc f
+    = Trivia Text
+    | Text Text
+    | Line Line
+    -- | Group predoc indicates either all or none of the spaces in predoc
+    -- and not in a subgroup should be converted to line breaks.
+    | Group (f (Predoc f))
+    -- | Nest n predoc indicates all line start in predoc should be indented by
+    -- n more spaces than the surroundings.
+    | Nest Int (f (Predoc f))
+
+deriving instance Show (Predoc Tree)
+deriving instance Show (Predoc [])
+
+type Doc = Tree (Predoc Tree)
+type DocList = [Predoc []]
+
+class Pretty a where
+    pretty :: a -> Doc
+
+instance Pretty Text where
+    pretty = Leaf . Text
+
+instance Pretty String where
+    pretty = Leaf . Text . pack
+
+instance Semigroup (Tree a) where
+    left <> right = Node left right
+
+instance Monoid (Tree (Predoc Tree)) where
+    mempty = EmptyTree
+
+text :: Text -> Doc
+text = Leaf . Text
+
+trivia :: Text -> Doc
+trivia = Leaf . Trivia
+
+group :: Doc -> Doc
+group = Leaf . Group
+
+nest :: Int -> Doc -> Doc
+nest level = Leaf . Nest level
+
+softbreak :: Doc
+softbreak = Leaf (Line Softbreak)
+
+break :: Doc
+break = Leaf (Line Break)
+
+softline :: Doc
+softline = Leaf (Line Softline)
+
+space :: Doc
+space = Leaf (Line Space)
+
+hardline :: Doc
+hardline = Leaf (Line Hardline)
+
+emptyline :: Doc
+emptyline = Leaf (Line Emptyline)
+
+hcat :: [Doc] -> Doc
+hcat = mconcat
+
+flatten :: Doc -> DocList
+flatten = go []
+    where go xs (Node x y)           = go (go xs y) x
+          go xs EmptyTree            = xs
+          go xs (Leaf (Group tree))  = Group (go [] tree) : xs
+          go xs (Leaf (Nest l tree)) = Nest l (go [] tree) : xs
+          go xs (Leaf (Line line))   = Line line : xs
+          go xs (Leaf (Text t))      = Text t : xs
+          go xs (Leaf (Trivia t))    = Trivia t : xs
+
+isLine :: Predoc f -> Bool
+isLine (Line _) = True
+isLine _        = False
+
+spanEnd :: (a -> Bool) -> [a] -> ([a], [a])
+spanEnd p = fmap reverse . span p . reverse
+
+-- | Lines before a Nest should be moved inside, breaks at the end of a nest
+-- should be moved outside. Lines at the start or end of a group should be
+-- moved outside.
+
```

Comment:
👾

Reply:
I improved the documentation, let me know if it's still unclear.